### DISCUSSION TITLE
[Guide] Emoji vs Emote in JDA

### DISCUSSION THREAD

**Discussion by Andre601** [discord-jda/JDA#1553] [General]
[emote]: https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/entities/Emote.html
[emoji]: https://emojipedia.org/

[message]: https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/entities/Message.html
[reaction]: https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/entities/MessageReaction.html
[addreaction]: https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/entities/Message.html#addReaction(net.dv8tion.jda.api.entities.Emote)
[getemote]: https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/entities/Guild.html#getEmoteById(long)
[getasmention]: https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/entities/Emote.html#getAsMention()

[check]: https://emojipedia.org/check-mark-button/
[zero]: https://emojipedia.org/keycap-digit-zero/
[one]: https://emojipedia.org/keycap-digit-one/
[nine]: https://emojipedia.org/keycap-digit-nine/

## About
JDA supports [Emotes][emote] and [Emojis][emoji] in various places like messages or reactions.
It can however get a bit confusing on either *how* you use it or *where* you can use what format.

## What are Emotes and what are Emojis?
As [Emotes][emote] does JDA usually refer to custom Emojis by Discord.  
Unlike Emojis, which are Unicode are Emotes in a different format.

### Formats (Emojis)
As mentioned before are Emojis Unicode and depending on where you use them can you use two different types of formats:

- In [Messages][message] can you use the raw Unicode-Format such as [`\u2705` (✅)][check]
- In [Reactions][reaction] ([`Message#addReaction(String)`][addreaction]) can you use the raw Unicode-Format but also the Codepoint format (`U+2705`).
  It's worth mentioning, that the Codepoint representation may have a different format than the Unicode one as it usually is only a single piece while Unicode can be made out of several parts.

#### Pro-tip: Number Emojis (1️⃣)
If you want to use any of the Numbered Emojis ([`:zero:` (0️⃣)][zero] to [`:nine:` (9️⃣)][nine]) can you save some work by simplifying the Emoji's Unicode format.
The emoji itself is put together from 2 separate Unicodes: The first one is the number, and the second one the square surrounding it. There is usually also a "Regional Indicator" put in-between those two but it can often be omitted without problems.

So if you f.e. want to display [1️⃣][one] can the full code (`\u0031\u20E3`) be simplified to `1\u20E3` since the first part itself is the number.
This also allows us to have a much simpler system to f.e. add multiple number reactions to a message:

```java
// Adds 0, 1, 2, 3, 4, 5, 6, 7, 8 and 9 Emojis.
private void addReactions(Message msg)
{
    for (int i = 0; i < 9; i++)
    {
        msg.addReaction(i + "\u20E3").queue();
    }
}
```

### Formats (Emotes)
Since [Emotes][emote] are Discord's custom Emojis do they also follow their own unique pattern to use in [messages][message] and [reactions][reaction].

- `<:_name_:_id_>` is used for displaying normal, non-animated Emotes
- `<a:_name_:_id_>` is used for displaying animated Emotes

Something that is worth pointing out is, that `_name_` can be any text you like as long as the `_id_` is a valid Emote ID.

IDs of [Emotes][emote] cannot be retrieved the normal way as you can do with IDs of channels, messages, etc.
Instead what you need to do is escape the [Emote] by prefixing it with a backslash (`\:emoteName:`) in the Textarea of the Discord Client to reveal the "raw" format. From there can you then take the ID from the corresponding section.

#### Usage in Reactions
The [`addReaction`][addreaction] method of JDA has an option that allows you to provide the Emote through a simple String instead of the [Emote] Object itself.

This allows you to essentially skip getting the [Emote] ([`Guild#getEmoteById(long)`][getemote]) and instead directly add the reaction.

In this quick example do we use the JDA Emote (`<:jda:230988580904763393>`)
```java
// With Emote Object
private void addReaction(Guild guild, Message msg)
{
    // Get the Emote from the Guild
    Emote emote = guild.getEmoteById(230988580904763393L);

    msg.addReaction(emote).queue();
}

// With String
private void addReaction(Message msg)
{
    msg.addReaction("jda:230988580904763393").queue();
}
```
Animated Emotes need to be prefixed with `a:` similar to how they're prefixed in their raw format.

## Summary
Here is a quick summary of what you should've learned now:

- [Emojis][emoji] are Unicode and [Emotes][emote] have their own unique format.
- [Messages][message] only allow the usage of the Unicode format (`\u<code>`) for [Emojis][emoji] and the "raw" format of [Emotes][emote] (Can be achieved using [`Emote#getAsMention()`][getasmention])
- [Reactions][reaction] accept Unicode, Codepoint representation, [Emote Object][emote], or the `_name_:_id_`/`a:_name_:_id_` format (Obviously replace `_name_` and `_id_` with the actual values).

### Can I use `:name:` instead of the Unicode?
No.
The `:name:` depends on the OS you use and can therefore be different while the Unicode is always the same. Because of this does JDA not have any direct support for using the `:name:` pattern to use Emojis in Messages or reactions.
It may work but it's not guaranteed.

You *can* add this support to your bot yourself by using an Emoji library for Java tho.

### Can my Bot use Emotes from other Guilds?
Yes, as long as the bot is on this Guild.

The only Entity allowed to use Emotes without this requirement is Webhooks. Those can use any Emote as long as the pattern is correct.

### Can I escape an Emoji to get its raw Unicode?
Not directly.
Escaping the Emoji (Prefixing it with a backslash) will only reveal the actual Emoji used. Discord uses Twemoji which makes them look the same across all platforms (with the exception for reactions) and escaping the Emoji will show the actual Emoji, which will look different per OS (See the [Emojipedia][emoji] for the different looks)

Similar to the first question can you use an Emoji Library to get the Unicode from an Emoji in a message, or look it up on [Emojipedia][emoji].

**Comment by ggiande** [comment]
I found through the community that a great use of the following implementation works better off than making the Emote object. 
![image](https://user-images.githubusercontent.com/44848977/115827193-cf2fd980-a3c0-11eb-8a6e-cd0e202034c9.png)

Please make it apparent to run that command inside of the Discord Server/Guild.

**Reply by Andre601** [reply]
What?

**Reply by ggiande** [reply]
That specific command does not need to run in Java; a developer can use it in the discord server. State that as an option.

**Reply by Andre601** [reply]
I never said this to be run "in Java"?
Like where do you even get that from? This section only says that you need to escape the Emote to get its raw format. At no point does it mention that you need to do this in Java, so you essentially assume something wrong here.

**Reply by ggiande** [reply]
You surround the guide with Java code blocks and make no mention of what can be done through Discord. I think it would be great if anyone who comes across the guide does not get confused by knowing what can be accomplished by Discord...

**Reply by Andre601** [reply]
I leave the section as is as I really doubt people are so stupid to not realize that you do this stuff in the Client itself.
So drop this already as it wouldn't bring any benefit at all to this guide.
